From e5246280ffc54faf9a43c269c24d4e259abbbf0a Mon Sep 17 00:00:00 2001
From: David Andrey <david.andrey@netmodule.com>
Date: Sat, 8 Dec 2012 10:12:13 +0100
Subject: [PATCH 05/21] ARM: zynq: I2C: check for transfert size

Check that the transfert size in I2C read do not
exceed the hardware capability

Signed-off-by:David Andrey david.andrey@netmodule.com
---
 drivers/i2c/zynq_i2c.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)
 mode change 100644 => 100755 drivers/i2c/zynq_i2c.c

diff --git a/drivers/i2c/zynq_i2c.c b/drivers/i2c/zynq_i2c.c
old mode 100644
new mode 100755
index 084cd27..e22d432
--- a/drivers/i2c/zynq_i2c.c
+++ b/drivers/i2c/zynq_i2c.c
@@ -118,7 +118,8 @@ struct zynq_i2c_registers {
 #error You must select CONFIG_ZYNQ_I2C_CTLR_0 or CONFIG_ZYNQ_I2C_CTLR_1
 #endif
 
-#define ZYNQ_I2C_FIFO_DEPTH 16
+#define ZYNQ_I2C_FIFO_DEPTH			16
+#define ZYNQ_I2C_TRANSFERT_SIZE_MAX	(255)	/* Controller transfer limit */
 
 static struct zynq_i2c_registers *zynq_i2c =
 	(struct zynq_i2c_registers *) ZYNQ_I2C_BASE;
@@ -213,6 +214,11 @@ int i2c_read(u8 dev, uint addr, int alen, u8 *data, int length)
 	u32 i=0;
 	u8 *cur_data = data;
 
+	/* check the hardware can handle the requested bytes */
+	if ((length < 0) || (length > ZYNQ_I2C_TRANSFERT_SIZE_MAX)) {
+		return EINVAL;
+	}
+
 	/* Write the register address */
 	Xsetbits_le32(&zynq_i2c->control, ZYNQ_I2C_CONTROL_CLR_FIFO |
 		ZYNQ_I2C_CONTROL_HOLD);
-- 
1.7.4.1

